velox/functions/sparksql/fuzzer/SparkQueryRunner.h

```diff
@@ -0,0 +1,105 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "grpc++/channel.h"
+#include "grpc++/client_context.h"
+#include "velox/common/memory/Memory.h"
+#include "velox/exec/fuzzer/ReferenceQueryRunner.h"
+#include "velox/functions/sparksql/fuzzer/proto/spark/connect/base.grpc.pb.h"
+#include "velox/vector/ComplexVector.h"
+
+namespace facebook::velox::functions::sparksql::fuzzer {
+
+/// Query runner that uses Spark as a reference database. It converts Velox
+/// query plan to Spark SQL and executes it in Spark. The results are returned
+/// as Velox compatible format.
+class SparkQueryRunner : public velox::exec::test::ReferenceQueryRunner {
+ public:
+  /// @param coordinatorUri Spark connect server endpoint, e.g. localhost:15002.
+  SparkQueryRunner(const std::string& coordinatorUri);
+
+  /// Converts Velox query plan to Spark SQL. Supports Values -> Aggregation.
+  /// Values node is converted into reading from 'tmp' table.
+  /// @return std::nullopt for unsupported cases.
+  std::optional<std::string> toSql(
+      const velox::core::PlanNodePtr& plan) override;
+
+  std::multiset<std::vector<velox::variant>> execute(
+      const std::string& sql,
+      const std::vector<velox::RowVectorPtr>& input,
+      const velox::RowTypePtr& resultType) override;
+
+  std::multiset<std::vector<velox::variant>> execute(
+      const std::string& sql,
+      const std::vector<RowVectorPtr>& probeInput,
+      const std::vector<RowVectorPtr>& buildInput,
+      const RowTypePtr& resultType) override {
+    VELOX_NYI("SparkQueryRunner does not support join node.");
+  }
+
+  bool supportsVeloxVectorResults() const override {
+    return true;
+  }
+
+  std::vector<velox::RowVectorPtr> executeVector(
+      const std::string& sql,
+      const std::vector<velox::RowVectorPtr>& input,
+      const velox::RowTypePtr& resultType) override;
+
+  std::vector<velox::RowVectorPtr> execute(const std::string& sql) override;
+
+ private:
+  velox::memory::MemoryPool* pool() {
+    return pool_.get();
+  }
+
+  velox::memory::MemoryPool* rootPool() {
+    return rootPool_.get();
+  }
+
+  // Reads the arrow IPC-format string data with arrow IPC reader and convert
+  // them into Velox RowVectors.
+  std::vector<velox::RowVectorPtr> readArrowData(const std::string& data);
+
+  // Returns false for unsupported aggregation in Spark.
+  bool supported(
+      const std::shared_ptr<const core::AggregationNode>& aggregationNode);
+
+  std::optional<std::string> toSql(
+      const std::shared_ptr<const velox::core::AggregationNode>&
+          aggregationNode);
+
+  std::optional<std::string> toSql(
+      const std::shared_ptr<const core::ProjectNode>& projectNode);
+
+  google::protobuf::Arena arena_;
+  // Used to make gRPC calls to the SparkConnectService.
+  std::unique_ptr<spark::connect::SparkConnectService::Stub> stub_;
+  std::shared_ptr<velox::memory::MemoryPool> rootPool_{
```

Comment:
I wonder if it makes sense to change the API to pass memory pool as an argument to 'execute'. CC: @xiaoxmeng @kgpai @yanngyoung

Reply:
Opened https://github.com/facebookincubator/velox/pull/10568 to make the API accept a memory pool parameter. Thanks.